File: src/Dwarf/structs.jl
struct DIE
  tag::Tag
  attrs::Vector{Pair{Attr,Any}}
  children::Vector{DIE}
end

DIE(tag::Tag, attrs = []) = DIE(tag, attrs, [])

struct Abbrev
  tag::Tag
  attrs::Vector{Pair{Attr,Form}}
  children::Bool
end

Base.hash(x::Abbrev, h::UInt) = Base.hash((x.tag, x.attrs, x.children), 0x0aa35d1caf09e6b0)

Base.:(==)(a::Abbrev, b::Abbrev) = (a.tag, a.attrs, a.children) == (b.tag, b.attrs, b.children)

const attrforms =
  Dict(AT_high_pc => (FORM_addr, UInt32),
       AT_low_pc  => (FORM_addr, UInt32),
       AT_stmt_list => (FORM_sec_offset, UInt32))

const byteforms =
  Dict(1 => FORM_data1,
       2 => FORM_data2,
       4 => FORM_data4,
       8 => FORM_data8)

form(x::Union{AbstractString,Symbol}) = FORM_string
form(x::Union{Enum,Integer}) = byteforms[sizeof(x)]
form(x::Bool) = FORM_flag

function form(attr, v)
  haskey(attrforms, attr) || return form(v)
  f, T = attrforms[attr]
  @assert v isa T
  return f
end

function abbrev(d::DIE)
  Abbrev(d.tag, [k => form(k, v) for (k, v) in d.attrs], !isempty(d.children))
end

function abbrevs!(die::DIE, as::Set{Abbrev})
  push!(as, abbrev(die))
  foreach(d -> abbrevs!(d, as), die.children)
  return as
end

abbrevs(die::DIE) = collect(abbrevs!(die, Set{Abbrev}()))

struct LineTable
  lines::Vector{Pair{UInt32,Union{Source,Nothing}}}
end

offset(lt::LineTable, δ) = LineTable([o-δ => s for (o, s) in lt.lines])

File: src/dwarf/structs.ts
import { HashSet } from '../utils/map'
import { Tag, Attr, Form } from './enums'

export {
  Source, Def, Frame, Stack, LineInfo, Function, Value, DIE, Abbrev, LineTable, offset, abbrev, abbrevs
}

interface Source {
  readonly file: string
  readonly line: number
  readonly col: number
}

interface Def {
  readonly name: string
  readonly source: Source | undefined
  readonly trampoline: boolean
}

function Def(name: string, source?: Source, trampoline: boolean = false): Def {
  return { name, source, trampoline }
}

type Frame = [Def, Source | undefined]
type Stack = Frame[]

interface LineInfo {
  readonly src: Stack
  readonly bp: boolean
}

function LineInfo(src: Stack, bp: boolean = false): LineInfo {
  return { src, bp }
}

interface Function {
  def: Def
  range: [number, number]
  inlines: [Function, Source | undefined][]
}

type Value =
  | [Form.addr, number]
  | [Form.sec_offset, number]
  | [Form.data1, number]
  | [Form.data2, number]
  | [Form.data4, number]
  | [Form.data8, bigint]
  | [Form.string, string]
  | [Form.flag, boolean]

class DIE {
  constructor(readonly tag: Tag, readonly attrs: [Attr, Value][], readonly children: DIE[]) { }
}

class Abbrev {
  constructor(readonly tag: Tag, readonly attrs: [Attr, Form][], readonly children: boolean) { }
}

class LineTable {
  constructor(readonly lines: [number, LineInfo][]) { }
}

function offset(lt: LineTable, δ: number): LineTable {
  return new LineTable(lt.lines.map(([o, s]) => [o - δ, s]))
}

function abbrev(d: DIE): Abbrev {
  return new Abbrev(d.tag, d.attrs.map(([a, [f]]) => [a, f]), d.children.length > 0)
}

function collect(die: DIE, as: HashSet<Abbrev>): HashSet<Abbrev> {
  as.add(abbrev(die))
  for (const c of die.children) collect(c, as)
  return as
}

function abbrevs(die: DIE): Abbrev[] {
  return [...collect(die, new HashSet())]
}
